pkg/tar/tar_test.go

```diff
@@ -185,23 +187,60 @@ func TestCreateTar(t *testing.T) {
 	verifyTarFile(t, tarFile, testFiles, testLinks)
 }
 
-func createTestTar(files []fileDesc, writer io.Writer) {
+func createTestTar(files []fileDesc, writer io.Writer, t *testing.T) {
 	tw := tar.NewWriter(writer)
+	defer tw.Close()
 	for _, fd := range files {
-		contentBytes := []byte(fd.content)
-		hdr := &tar.Header{
-			Name:       fd.name,
-			Mode:       int64(fd.mode),
-			Size:       int64(len(contentBytes)),
-			Typeflag:   tar.TypeReg,
-			AccessTime: time.Now(),
-			ModTime:    fd.modifiedDate,
-			ChangeTime: fd.modifiedDate,
+		if isSymLink(fd.mode) {
+			if err := addSymLink(tw, &fd, t); err != nil {
+				t.Fatalf(
+					"Unable to add symbolic link %s (that points to %s) to archive: %v",
+					fd.name, fd.target, err,
+				)
+			}
+		} else {
+			addRegularFile(tw, &fd)
 		}
-		tw.WriteHeader(hdr)
-		tw.Write(contentBytes)
 	}
-	tw.Close()
+}
+
+func addRegularFile(tw *tar.Writer, fd *fileDesc) {
+	contentBytes := []byte(fd.content)
+	hdr := &tar.Header{
+		Name:       fd.name,
+		Mode:       int64(fd.mode),
+		Size:       int64(len(contentBytes)),
+		Typeflag:   tar.TypeReg,
+		AccessTime: time.Now(),
+		ModTime:    fd.modifiedDate,
+		ChangeTime: fd.modifiedDate,
+	}
+	tw.WriteHeader(hdr)
+	tw.Write(contentBytes)
+}
+
+func addSymLink(tw *tar.Writer, fd *fileDesc, t *testing.T) error {
+	if len(fd.target) == 0 {
+		return fmt.Errorf("Link %s must points to somewhere, but it points to nothing", fd.name)
+	}
+
+	hdr := &tar.Header{
+		Name:     fd.name,
+		Linkname: fd.target,
+		Mode:     int64(fd.mode & os.ModePerm),
+		Typeflag: tar.TypeSymlink,
+		ModTime:  fd.modifiedDate,
+	}
+
+	if err := tw.WriteHeader(hdr); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
@php-coder this pattern applies when the function signatures are the same. E.g.:

```go
func foo(x, y int) (int, error) {
	if y == 1 {
		return 0, fmt.Errorf("bad value 1")
	}
	return x + y, nil
}

func bar(z int) (int, error) {
	if z == 2 {
		return 0, fmt.Errorf("bad value 2")
	}
	// simply return _foo_, because it has
	// the same signature of _bar_
	return foo(1, z)
}
```